### PR TITLE
capi-kubevirt: Add E2E test

### PR DESCRIPTION
**Pull request by nunnatsa**
Signed-off-by: Nahshon Unna-Tsameret <nunnatsa@redhat.com>

### CODE REVIEW

**Comment by nirarg** [comment]
@nunnatsa as discussed, `openshift/cluster-api-actuator-pkg` shouldn't be used here
It doesn't test the `cluster-api` cluster

File: ci-operator/config/openshift/cluster-api-provider-kubevirt/openshift-cluster-api-provider-kubevirt-main.yaml [hunk 1]
```diff
@@ -78,6 +101,238 @@ tests:
...
+        [mysection]
+        keyUsage = keyEncipherment, keyCertSign, digitalSignature
+        EOF
+
+        openssl req -x509 -newkey rsa:4096 -sha256 -days 14 -nodes -keyout capi-webhooks.key -out capi-webhooks.crt -subj "/OU=openshift/CN=capi-webhooks" -extensions mysection -config ssl.cfg
+
+        openssl x509 -in capi-webhooks.crt -out capi-webhooks.pem  -outform PEM
+        openssl rsa -in capi-webhooks.key -outform pem -out capi-webhooks.key.pem
+
+        ls -la
```
Comment (davidvossel) [inline ci-operator/config/openshift/cluster-api-provider-kubevirt/openshift-cluster-api-provider-kubevirt-main.yaml:212]: what's the reasoning for needing to generate your own certs like this? Is it not practical to install and use certmanager like we do in the upstream e2e tests?

Comment (nunnatsa) [reply ci-operator/config/openshift/cluster-api-provider-kubevirt/openshift-cluster-api-provider-kubevirt-main.yaml:212]: I tried that using `oc kustomize default` but I've got "wrong certificate" error or something like that.

File: ci-operator/config/openshift/cluster-api-provider-kubevirt/openshift-cluster-api-provider-kubevirt-main.yaml [hunk 2]
```diff
@@ -78,6 +101,242 @@ tests:
...
+        requests:
+          cpu: 500m
+          memory: 1Gi
+    - as: create-and-test-cluster
+      commands: |-
+        set -x
+        export NODE_VM_IMAGE_TEMPLATE=quay.io/kubevirtci/fedora-kubeadm:35
+        export IMAGE_REPO=k8s.gcr.io
+        export CRI_PATH="/var/run/crio/crio.sock"
+        clusterctl generate cluster kvcluster --kubernetes-version v1.21.0 --control-plane-machine-count=1 --worker-machine-count=1 --from templates/cluster-template.yaml | oc apply -f -
```
Comment (davidvossel) [inline ci-operator/config/openshift/cluster-api-provider-kubevirt/openshift-cluster-api-provider-kubevirt-main.yaml:328]: I think we need to wait for the capk operator to come online. The CI lane fails because the webhook isn't serving yet.

so, wait for the operator deployment. probably something like this...

```
oc wait -n system --for=condition=Available=true deployment/capi-kubevirt-operator --timeout=10m
```


```
Error from server (InternalError): error when creating "STDIN": Internal error occurred: failed calling webhook "validation.kubevirtmachinetemplate.infrastructure.cluster.x-k8s.io": Post "https://webhook-service.system.svc:443/validate-infrastructure-cluster-x-k8s-io-v1alpha1-kubevirtmachinetemplate?timeout=10s": no endpoints available for service "webhook-service"
Error from server (InternalError): error when creating "STDIN": Internal error occurred: failed calling webhook "validation.kubevirtmachinetemplate.infrastructure.cluster.x-k8s.io": Post "https://webhook-service.system.svc:443/validate-infrastructure-cluster-x-k8s-io-v1alpha1-kubevirtmachinetemplate?timeout=10s": no endpoints available for service "webhook-service"
```

File: ci-operator/config/openshift/cluster-api-provider-kubevirt/openshift-cluster-api-provider-kubevirt-main.yaml [hunk 3]
```diff
@@ -78,6 +101,244 @@ tests:
...
+        oc adm policy add-scc-to-user privileged system:serviceaccount:capk-system:capk-manager
+
+        export MANIFEST_IMG=${CAPK_IMAGE%:*}
+        export MANIFEST_TAG=${CAPK_IMAGE/*:}
+        GOFLAGS="-mod=readonly" make generate-manifests
+
+        sed -i -E "s|(^ +- image: ).*$|\1${CAPK_IMAGE}|" config/default/manager_image_patch.yaml
+
+        oc kustomize config/default | oc apply -f -
+        sleep 60
```
Comment (nirarg) [inline ci-operator/config/openshift/cluster-api-provider-kubevirt/openshift-cluster-api-provider-kubevirt-main.yaml:211]: Why is this needed, you have the `oc wait` after

File: ci-operator/config/openshift/cluster-api-provider-kubevirt/openshift-cluster-api-provider-kubevirt-main.yaml [hunk 4]
```diff
@@ -78,6 +101,244 @@ tests:
...
+        requests:
+          cpu: 500m
+          memory: 1Gi
+    - as: validate-cluster
+      commands: |-
+        set -x
+        export CLUSTER_NAME=tenant-capk
+        export CLUSTER_NS=capk-system
+
+        sleep 60
```
Comment (nirarg) [inline ci-operator/config/openshift/cluster-api-provider-kubevirt/openshift-cluster-api-provider-kubevirt-main.yaml:249]: Why need sleep here, you have wait next line

File: ci-operator/config/openshift/cluster-api-provider-kubevirt/openshift-cluster-api-provider-kubevirt-main.yaml [hunk 5]
```diff
@@ -78,6 +101,244 @@ tests:
...
+        if [[ ${NODES_ARE_UP} == "false" ]]; then
+          echo "Should have two working nodes by now; aborting..."
+          exit 1
+        fi
+
+        oc get nodes -o wide
+
+        oc config set-context $(oc config current-context) --namespace=default
+
+        # create a workload on the tenant cluster
```
Comment (nirarg) [inline ci-operator/config/openshift/cluster-api-provider-kubevirt/openshift-cluster-api-provider-kubevirt-main.yaml:332]: I think this shouldn't be done here, maybe better to add the test in capk repository, also to be used upstream?
Need to be executed in another step

File: ci-operator/config/openshift/cluster-api-provider-kubevirt/openshift-cluster-api-provider-kubevirt-main.yaml (hunk 3 above)
Comment (nunnatsa) [reply ci-operator/config/openshift/cluster-api-provider-kubevirt/openshift-cluster-api-provider-kubevirt-main.yaml:211]: Because I want to make sure that resource exist before waiting for its conditions. The wait will fail if the resource does not exist.

File: ci-operator/config/openshift/cluster-api-provider-kubevirt/openshift-cluster-api-provider-kubevirt-main.yaml (hunk 4 above)
Comment (nunnatsa) [reply ci-operator/config/openshift/cluster-api-provider-kubevirt/openshift-cluster-api-provider-kubevirt-main.yaml:249]: Because I want to make sure that resource exist before waiting for its conditions. The wait will fail if the resource does not exist.

**Review by davidvossel — COMMENTED** [review]
Is there a reason why we can't run the capk e2e test suite instead of all the bash validatino logic?

For example, I'd expect these steps

1. install dependencies (cnv + capi + cert manager)
2. install capk from source
3. run `make e2e` from source to execute e2e tests against the cluster that's been setup.


You already have logic for steps 1 and 2 that can be re-used. Step 3 using `make e2e` seems like it could replace everything from the `clusterctl generate cluster` and beyond.

**Review by davidvossel — COMMENTED** [review]
/lgtm
/approve

File: core-services/image-mirroring/openshift/mapping_origin_4_11 [hunk 6]
```diff
@@ -37,6 +37,7 @@ registry.ci.openshift.org/origin/4.11:baremetal-operator quay.io/openshift/origi
 registry.ci.openshift.org/origin/4.11:baremetal-runtimecfg quay.io/openshift/origin-baremetal-runtimecfg:4.11 quay.io/openshift/origin-baremetal-runtimecfg:4.11.0 quay.io/openshift/origin-baremetal-runtimecfg:latest
 registry.ci.openshift.org/origin/4.11:base quay.io/openshift/origin-base:4.11 quay.io/openshift/origin-base:4.11.0 quay.io/openshift/origin-base:latest
 registry.ci.openshift.org/origin/4.11:build-test-qemu-img quay.io/openshift/origin-build-test-qemu-img:4.11 quay.io/openshift/origin-build-test-qemu-img:4.11.0 quay.io/openshift/origin-build-test-qemu-img:latest
+registry.ci.openshift.org/origin/4.11:capk-oc-bin-image quay.io/openshift/origin-capk-oc-bin-image:4.11 quay.io/openshift/origin-capk-oc-bin-image:4.11.0 quay.io/openshift/origin-capk-oc-bin-image:latest
```
[the remark this answers is not in the document]
Comment (smarterclayton) [reply core-services/image-mirroring/openshift/mapping_origin_4_11:40]: @bparees

[the remark this answers is not in the document]
Comment (nunnatsa) [reply core-services/image-mirroring/openshift/mapping_origin_4_11:40]: > What is `capk`? I would assume this is not an image we publish (in which case it shouldn't show up)

@smarterclayton - We do need it for the hypershift project, so it could support kubevirt as a cluster api provider.

[the remark this answers is not in the document]
Comment (nunnatsa) [reply core-services/image-mirroring/openshift/mapping_origin_4_11]: Removed. Thanks @smarterclayton

**Review by davidvossel — APPROVED** [review]
/lgtm

Thanks!
